### DISCUSSION TITLE
API Breaking Changes, createTextNote

### DISCUSSION THREAD

**Discussion by MorganeLeCaignec** [TriliumNext/Trilium#4235] [Q&A] [answered]
So I've been trying to update my widgets, and there's a bunch of methods that used to be in the API that are just gone and I can't even find a reference to them.
Would have been nice to still have them written somewhere so I could know what to replace them with.

I used to use `api.createNoteLink()` for example which would return the actual HTML Element that I could then modify and append somewhere in my DOM. Now the Doc recommends using `api.createLink()`, except for some reason `createLink()` has no return so I don't really what it does and what I'm supposed to use it for, create a link that I can never get and use? I'm probably missing something.

One in particular I have no idea how to replace is `api.createTextNote()` which used to allow creating a new note programmatically. I've looked at the frontend API doc but nothing seems to allow me to do this now.

**Comment by sottey** [comment, marked as answer]
Creating a new note requires updating the database, so this is not a frontend process. It is in the backend api, not the frontend api. Take a look at [createNewNote()](https://zadam.github.io/trilium/backend_api/BackendScriptApi.html#createNewNote) and [createNote()](https://zadam.github.io/trilium/backend_api/BackendScriptApi.html#createNote) in the backend api documentation

Hope that helps!
Sean

**Reply by MorganeLeCaignec** [reply]
Damn I could have sworn I looked but I must have not been looking at the right version, or the right API page. Shameful.
Thank you!

**Comment by zerebos** [comment]
`createLink` should return a jQuery element just like before, take a look here https://github.com/zadam/trilium/blob/f5ddabc04229a82d946107e7bdac6d32e2c767c1/src/public/app/services/link.js#L28

**Reply by MorganeLeCaignec** [reply]
Huh, you're correct, I should have looked at the source. I was referencing the Docs which didn't include a return, and I think the typing library I was trying (https://github.com/rauenzi/trilium-types) also didn't include the return type (probably because it's copied/based on the API Docs), so it was giving me warnings.

Thank you!

@rauenzi This is probably a bug. Although I'd say it's more of a case of the Docs being outdated. But your package is really useful!

**Reply by zerebos** [reply]
Dang I thought I had fixed that. I actually ended up going through and manually verifying most of the function signatures while making the types package. I even contributed many of the fixes back upstream to improve the API docs https://github.com/zadam/trilium/pull/4188

**Comment by zadam** [comment]
I have added jsDoc @returns for the `api.createLink()`. The reason the old call was deprecated was that this method is more general and can link to e.g. attachments, and not just notes.

There were some other breaking changes (removal of deprecated methods) in the past releases, but all should have a simple replacement.
